### PR TITLE
Feature/272 variable comparisons

### PR DESCRIPTION
**Pull request by aannleax**
Implements comparison operations between variables, i.e. you can now write `?X < ?Y` in the rule body. It should now also handle multiple comparisons correctly (before, writing `?X < 4` and `?X < 8` resulted in the first restriction being ignored).

This functionality has been implemented by extending `ColumnScanEqualValue` which has now been renamed to `ColumnScanRestrictValues` (and similar for the associated trie scan).

### CODE REVIEW

File: nemo-physical/src/columnar/operations/columnscan_restrict_values.rs [hunk 1]
```diff
@@ -0,0 +1,355 @@
+use super::super::traits::columnscan::{ColumnScan, ColumnScanCell};
+use crate::datatypes::ColumnDataType;
+use std::{fmt::Debug, ops::Range};
+
+/// Concrete value a interval bound can take.
```
Comment (mmarx) [inline nemo-physical/src/columnar/operations/columnscan_restrict_values.rs:5]: `an interval`

File: nemo-physical/src/columnar/operations/columnscan_restrict_values.rs [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+    T: Clone,
+{
+    /// Return the column index this value refers to.
+    /// Returns None if this is a constant.
+    pub fn column_index(&self) -> Option<usize> {
+        if let Self::Column(index) = self {
+            Some(*index)
+        } else {
+            None
+        }
```
Comment (mmarx) [inline nemo-physical/src/columnar/operations/columnscan_restrict_values.rs:28]: Please use an explicit `match`, so we will not silently do the wrong thing when a new variant is added to the enum.

File: nemo-physical/src/columnar/operations/columnscan_restrict_values.rs [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+    }
+
+    /// Return a mutable reference to the column index this value refers to.
+    /// Returns None if this is a constant.
+    pub fn column_index_mut(&mut self) -> Option<&mut usize> {
+        if let Self::Column(index) = self {
+            Some(index)
+        } else {
+            None
+        }
```
Comment (mmarx) [inline nemo-physical/src/columnar/operations/columnscan_restrict_values.rs:38]: see above

File: nemo-physical/src/tabular/operations/triescan_select.rs [hunk 4]
```diff
@@ -224,63 +227,108 @@ impl<'a> TrieScanSelectValue<'a> {
...
+                                            "Expected a column scan of type {}",
+                                            stringify!($variant)
+                                        );
+                                    };
+
+                                    scans_restriction.push(referenced_scan);
+
+                                    map_len
+                                }
+                            };
```
Comment (mmarx) [inline nemo-physical/src/tabular/operations/triescan_select.rs:311]: Use `column_map.entry(*bound_index).or_insert_with(|| …)`

**Review by mmarx — CHANGES_REQUESTED** [review]
Just some small comments, looks good otherwise!

File: nemo-physical/src/util.rs [hunk 5]
```diff
@@ -1,6 +1,6 @@
 //! This module collects miscellaneous functionality.
 
-pub mod interval;
+// pub mod interval;
```
[the remark this answers is not in the document]
Comment (aannleax) [reply nemo-physical/src/util.rs:3]: Code there is not used anymore. I'll leave the file but remove the commented out module statement.

File: nemo-physical/src/util.rs [hunk 6]
```diff
@@ -1,6 +1,5 @@
 //! This module collects miscellaneous functionality.
 
-pub mod interval;
```
[the remark this answers is not in the document]
Comment (mmarx) [reply nemo-physical/src/util.rs:3]: Still better than having dead code that is not compiled, so we will never know when it breaks ;)

**Review by mmarx — APPROVED** [review]

**Pull request merged**